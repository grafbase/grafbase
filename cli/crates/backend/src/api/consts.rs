use std::sync::OnceLock;

pub const CREDENTIALS_FILE: &str = "credentials.json";
pub const AUTH_URL: &str = "https://app.grafbase.com/auth/cli";
pub const PACKAGE_JSON: &str = "package.json";
pub const TAR_CONTENT_TYPE: &str = "application/x-tar";
pub const GRAFBASE_ACCESS_TOKEN_ENV_VAR: &str = "GRAFBASE_ACCESS_TOKEN";

const API_URL: &str = "https://api.grafbase.com/graphql";
const DASHBOARD_URL: &str = "https://app.grafbase.com";

pub fn api_url() -> &'static str {
    static API_URL: OnceLock<String> = OnceLock::new();

    API_URL.get_or_init(|| match std::env::var("GRAFBASE_API_URL").ok() {
        Some(url) => url,
        None => self::API_URL.to_string(),
    })
}

pub fn dashboard_url() -> &'static str {
    static DASHBOARD_URL: OnceLock<String> = OnceLock::new();

    DASHBOARD_URL.get_or_init(|| match std::env::var("GRAFBASE_DASHBOARD_URL").ok() {
        Some(url) => url,
        None => self::DASHBOARD_URL.to_string(),
    })
}
