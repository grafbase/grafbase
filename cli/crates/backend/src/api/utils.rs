use super::{consts::PROJECT_METADATA_FILE, errors::ApiError, types::ProjectMetadata};
use common::environment::Project;
use tokio::fs;

pub async fn project_linked() -> Result<Option<ProjectMetadata>, ApiError> {
    let project = Project::get();

    let project_metadata_file_path = project.dot_grafbase_directory_path.join(PROJECT_METADATA_FILE);

    if !fs::try_exists(&project_metadata_file_path)
        .await
        .map_err(ApiError::ReadProjectMetadataFile)?
    {
        return Ok(None);
    }

    let project_metadata_file = tokio::fs::read_to_string(project_metadata_file_path)
        .await
        .map_err(ApiError::ReadProjectMetadataFile)?;

    Ok(Some(
        serde_json::from_str(&project_metadata_file).map_err(|_| ApiError::CorruptProjectMetadataFile)?,
    ))
}

pub async fn has_project_linked() -> Result<bool, ApiError> {
    let project = Project::get();
    fs::try_exists(project.dot_grafbase_directory_path.join(PROJECT_METADATA_FILE))
        .await
        .map_err(ApiError::ReadProjectMetadataFile)
}
