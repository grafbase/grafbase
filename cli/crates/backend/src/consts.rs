use const_format::formatcp;

pub const DEFAULT_SCHEMA_FEDERATED: &str = include_str!("../assets/grafbase.federated.config.ts");
pub const DEFAULT_SCHEMA_SINGLE: &str = include_str!("../assets/grafbase.single.config.ts");
pub const DEFAULT_DOT_ENV: &str = include_str!("../assets/default.env");
pub const USER_AGENT: &str = formatcp!("Grafbase-CLI-{}", env!("CARGO_PKG_VERSION"));
