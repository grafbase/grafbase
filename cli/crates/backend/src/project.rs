use crate::consts::{DEFAULT_DOT_ENV, DEFAULT_SCHEMA_SDL, DEFAULT_SCHEMA_TS, USER_AGENT};
use crate::errors::BackendError;
use async_compression::tokio::bufread::GzipDecoder;
use async_tar::Archive;
use common::consts::{
    GRAFBASE_DIRECTORY_NAME, GRAFBASE_ENV_FILE_NAME, GRAFBASE_SCHEMA_FILE_NAME, GRAFBASE_SDK_PACKAGE_NAME,
    GRAFBASE_SDK_PACKAGE_VERSION, GRAFBASE_TS_CONFIG_FILE_NAME,
};
use common::environment::{self, Project};
use http_cache_reqwest::{CACacheManager, Cache, CacheMode, HttpCache};
use reqwest::{header, Client};
use reqwest_middleware::ClientBuilder;
use serde::Deserialize;
use std::fs;
use std::io::{Error as IoError, ErrorKind as IoErrorKind};
use std::iter::Iterator;
use std::path::{Path, PathBuf};
use std::{env, fmt};
use tokio_stream::StreamExt;
use tokio_util::compat::TokioAsyncReadCompatExt;
use tokio_util::io::StreamReader;
use url::Url;

#[derive(Debug, Clone, Copy)]
pub enum ConfigType {
    TypeScript,
    GraphQL,
}

impl AsRef<str> for ConfigType {
    fn as_ref(&self) -> &str {
        match self {
            ConfigType::TypeScript => "typescript",
            ConfigType::GraphQL => "graphql",
        }
    }
}

impl fmt::Display for ConfigType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ConfigType::TypeScript => f.write_str("TypeScript"),
            ConfigType::GraphQL => f.write_str("GraphQL"),
        }
    }
}

impl ConfigType {
    pub const VARIANTS: &'static [ConfigType] = &[Self::TypeScript, Self::GraphQL];
}

#[derive(Debug, Clone, Copy)]
pub enum Template<'a> {
    FromUrl(&'a str),
    FromDefault(ConfigType),
}

/// initializes a new project in the current or a new directory, optionally from a template
///
/// # Errors
///
/// ## General
///
/// - returns [`BackendError::ReadCurrentDirectory`] if the current directory could not be read
///
/// - returns [`BackendError::ProjectDirectoryExists`] if a named is passed and a directory with the same name already exists in the current directory
///
/// - returns [`BackendError::AlreadyAProject`] if there's already a grafbase/schema.graphql in the target
///
/// - returns [`BackendError::CreateGrafbaseDirectory`] if the grafbase directory could not be created
///
/// - returns [`BackendError::CreateProjectDirectory`] if the project directory could not be created
///
/// - returns [`BackendError::WriteSchema`] if the schema file could not be written
///
/// ## Templates
///
/// - returns [`BackendError::UnsupportedTemplateURL`] if a template URL is not supported
///
/// - returns [`BackendError::StartDownloadRepoArchive`] if a template URL is not supported (if the request could not be made)
///
/// - returns [`BackendError::DownloadRepoArchive`] if a repo tar could not be downloaded (on a non 200-299 status)
///
/// - returns [`BackendError::TemplateNotFound`] if no files matching the template path were extracted (excluding extraction errors)
///
/// - returns [`BackendError::MoveExtractedFiles`] if the extracted files from the template repository could not be moved
///
/// - returns [`BackendError::ReadArchiveEntries`] if the entries of the template repository archive could not be read
///
/// - returns [`BackendError::ExtractArchiveEntry`] if one of the entries of the template repository archive could not be extracted
///
/// - returns [`BackendError::CleanExtractedFiles`] if the files extracted from the template repository archive could not be cleaned
///
/// - returns [`BackendError::StartGetRepositoryInformation`] if the request to get the information for a repository could not be sent
///
/// - returns [`BackendError::GetRepositoryInformation`] if the request to get the information for a repository returned a non 200-299 status
///
/// - returns [`BackendError::ReadRepositoryInformation`] if the request to get the information for a repository returned a response that could not be parsed
pub fn init(name: Option<&str>, template: Template<'_>) -> Result<ConfigType, BackendError> {
    let project_path = to_project_path(name)?;
    let grafbase_path = project_path.join(GRAFBASE_DIRECTORY_NAME);

    if grafbase_path.exists() {
        return Err(BackendError::AlreadyAProject(grafbase_path));
    }

    let runtime = tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .build()
        .unwrap();

    runtime.block_on(async move {
        match template {
            Template::FromUrl(template) => {
                // as directory names cannot contain slashes, and URLs with no scheme or path cannot
                // be differentiated from a valid template name,
                // anything with a slash is treated as a URL
                if template.contains('/') {
                    if let Ok(repo_url) = Url::parse(template) {
                        match repo_url.host_str() {
                            Some("github.com") => handle_github_repo_url(&grafbase_path, &repo_url).await?,
                            _ => return Err(BackendError::UnsupportedTemplateURL(template.to_string())),
                        }
                    } else {
                        return Err(BackendError::MalformedTemplateURL(template.to_owned()));
                    }
                } else {
                    download_github_template(
                        &grafbase_path,
                        GitHubTemplate::Grafbase(GrafbaseGithubTemplate { path: template }),
                    )
                    .await?;
                }

                if std::fs::read_dir(&grafbase_path)
                    .expect("We must have a valid directory in this point.")
                    .any(|item| {
                        item.ok()
                            .and_then(|dir_entry| dir_entry.path().extension().map(|extension| extension == "ts"))
                            .unwrap_or_default()
                    })
                {
                    environment::add_dev_dependency_to_package_json(
                        &project_path,
                        GRAFBASE_SDK_PACKAGE_NAME,
                        GRAFBASE_SDK_PACKAGE_VERSION,
                    )
                    .map_err(BackendError::CommonError)?;
                    return Ok(ConfigType::TypeScript);
                }

                Ok(ConfigType::GraphQL)
            }
            Template::FromDefault(config_type) => {
                tokio::fs::create_dir_all(&grafbase_path)
                    .await
                    .map_err(BackendError::CreateGrafbaseDirectory)?;

                let dot_env_path = grafbase_path.join(GRAFBASE_ENV_FILE_NAME);

                let schema_write_result = match config_type {
                    ConfigType::TypeScript => {
                        let schema_path = grafbase_path.join(GRAFBASE_TS_CONFIG_FILE_NAME);

                        let add_sdk = environment::add_dev_dependency_to_package_json(
                            &project_path,
                            GRAFBASE_SDK_PACKAGE_NAME,
                            GRAFBASE_SDK_PACKAGE_VERSION,
                        )
                        .map_err(Into::into);

                        let write_schema = fs::write(schema_path, DEFAULT_SCHEMA_TS).map_err(BackendError::WriteSchema);

                        add_sdk.and(write_schema)
                    }
                    ConfigType::GraphQL => {
                        let schema_path = grafbase_path.join(GRAFBASE_SCHEMA_FILE_NAME);
                        fs::write(schema_path, DEFAULT_SCHEMA_SDL).map_err(BackendError::WriteSchema)
                    }
                };

                let dot_env_write_result = fs::write(dot_env_path, DEFAULT_DOT_ENV).map_err(BackendError::WriteSchema);

                if schema_write_result.is_err() || dot_env_write_result.is_err() {
                    tokio::fs::remove_dir_all(&grafbase_path)
                        .await
                        .map_err(BackendError::DeleteGrafbaseDirectory)?;
                }

                schema_write_result?;
                dot_env_write_result?;

                Ok(config_type)
            }
        }
    })
}

async fn handle_github_repo_url(grafbase_path: &Path, repo_url: &Url) -> Result<(), BackendError> {
    if let Some(mut segments) = repo_url.path_segments().map(Iterator::collect::<Vec<_>>) {
        // remove trailing slashes to prevent extra path parameters
        if segments.last() == Some(&"") {
            segments.pop();
        }

        // disallow empty path paramters other than the last
        if segments.contains(&"") {
            return Err(BackendError::UnsupportedTemplateURL(repo_url.to_string()));
        }

        match segments.len() {
            2 => {
                let org = &segments[0];

                let repo = &segments[1];

                let branch = get_default_branch(org, repo).await?;

                download_github_template(
                    grafbase_path,
                    GitHubTemplate::External(ExternalGitHubTemplate {
                        org,
                        repo,
                        branch: &branch,
                        path: None,
                    }),
                )
                .await
            }
            4.. if segments[2] == "tree" => {
                let org = &segments[0];

                let repo = &segments[1];

                let branch = &segments[3];

                let path = segments.get(4..).map(|path| path.join("/"));

                download_github_template(
                    grafbase_path,
                    GitHubTemplate::External(ExternalGitHubTemplate {
                        org,
                        repo,
                        path,
                        branch,
                    }),
                )
                .await
            }
            _ => Err(BackendError::UnsupportedTemplateURL(repo_url.to_string())),
        }
    } else {
        Err(BackendError::UnsupportedTemplateURL(repo_url.to_string()))
    }
}

#[derive(Deserialize)]
struct RepoInfo {
    default_branch: String,
}

async fn get_default_branch(org: &str, repo: &str) -> Result<String, BackendError> {
    let client = Client::new();

    let response = client
        .get(format!("https://api.github.com/repos/{org}/{repo}"))
        // api.github.com requires a user agent header to be present
        .header(header::USER_AGENT, USER_AGENT)
        .send()
        .await
        .map_err(|_| BackendError::StartGetRepositoryInformation(format!("{org}/{repo}")))?;

    if !response.status().is_success() {
        return Err(BackendError::GetRepositoryInformation(format!("{org}/{repo}")));
    }

    let repo_info = response
        .json::<RepoInfo>()
        .await
        .map_err(|_| BackendError::ReadRepositoryInformation(format!("{org}/{repo}")))?;

    Ok(repo_info.default_branch)
}

fn to_project_path(name: Option<&str>) -> Result<PathBuf, BackendError> {
    let current_dir = env::current_dir().map_err(|_| BackendError::ReadCurrentDirectory)?;
    match name {
        Some(name) => {
            let project_path = current_dir.join(name);
            if project_path.exists() {
                Err(BackendError::ProjectDirectoryExists(project_path))
            } else {
                Ok(project_path)
            }
        }
        None => Ok(current_dir),
    }
}

#[derive(Clone)]
struct ExternalGitHubTemplate<'a> {
    org: &'a str,
    repo: &'a str,
    path: Option<String>,
    branch: &'a str,
}

struct GrafbaseGithubTemplate<'a> {
    path: &'a str,
}

enum GitHubTemplate<'a> {
    Grafbase(GrafbaseGithubTemplate<'a>),
    External(ExternalGitHubTemplate<'a>),
}

impl<'a> GitHubTemplate<'a> {
    pub fn into_external_github_template(self) -> ExternalGitHubTemplate<'a> {
        match self {
            Self::Grafbase(GrafbaseGithubTemplate { path }) => ExternalGitHubTemplate {
                org: "grafbase",
                repo: "grafbase",
                path: Some(format!("templates/{path}")),
                branch: "main",
            },
            Self::External(template @ ExternalGitHubTemplate { .. }) => template,
        }
    }
}

async fn download_github_template(grafbase_path: &Path, template: GitHubTemplate<'_>) -> Result<(), BackendError> {
    let ExternalGitHubTemplate {
        org,
        repo,
        path,
        branch,
    } = template.into_external_github_template();

    let org_and_repo = format!("{org}/{repo}");

    let extraction_dir = PathBuf::from(format!("{repo}-{branch}"));

    let mut template_path: PathBuf = PathBuf::from(&extraction_dir);

    if let Some(path) = path {
        template_path.push(path);
    }

    template_path.push("grafbase");

    let extraction_result = stream_github_archive(grafbase_path, &org_and_repo, template_path, branch).await;

    if extraction_dir.exists() {
        tokio::fs::remove_dir_all(extraction_dir)
            .await
            .map_err(BackendError::CleanExtractedFiles)?;
    }

    extraction_result
}

async fn stream_github_archive<'a>(
    grafbase_path: &Path,
    org_and_repo: &'a str,
    template_path: PathBuf,
    branch: &'a str,
) -> Result<(), BackendError> {
    // not using the common environment since it's not initialized here
    // if the OS does not have a cache path or it is not UTF-8, we don't cache the download
    let cache_directory = dirs::cache_dir().and_then(|path| path.join("grafbase").to_str().map(ToOwned::to_owned));

    let mut client_builder = ClientBuilder::new(Client::new());

    if let Some(cache_directory) = cache_directory {
        client_builder = client_builder.with(Cache(HttpCache {
            mode: CacheMode::Default,
            manager: CACacheManager { path: cache_directory },
            options: None,
        }));
    }

    let client = client_builder.build();

    let tar_gz_response = client
        .get(format!("https://codeload.github.com/{org_and_repo}/tar.gz/{branch}"))
        .send()
        .await
        .map_err(|error| BackendError::StartDownloadRepoArchive(org_and_repo.to_owned(), error))?;

    if !tar_gz_response.status().is_success() {
        return Err(BackendError::DownloadRepoArchive(org_and_repo.to_owned()));
    }

    let tar_gz_stream = tar_gz_response
        .bytes_stream()
        .map(|result| result.map_err(|error| IoError::new(IoErrorKind::Other, error)));

    let tar_gz_reader = StreamReader::new(tar_gz_stream);
    let tar = GzipDecoder::new(tar_gz_reader);
    let archive = Archive::new(tar.compat());

    let mut entries = archive.entries().map_err(|_| BackendError::ReadArchiveEntries)?;

    while let Some(entry) = entries.next().await {
        let mut entry = entry.map_err(BackendError::ExtractArchiveEntry)?;

        if entry
            .path()
            .ok()
            .filter(|path| path.starts_with(&template_path))
            .is_some()
        {
            entry.unpack_in(".").await.map_err(BackendError::ExtractArchiveEntry)?;
        }
    }

    if !template_path.exists() {
        return Err(BackendError::TemplateNotFound);
    }

    let project_folder = grafbase_path.parent().expect("must exist");

    let named_project = !project_folder.exists();

    if named_project {
        tokio::fs::create_dir(project_folder)
            .await
            .map_err(BackendError::CreateProjectDirectory)?;
    }

    let rename_result = tokio::fs::rename(template_path, &grafbase_path)
        .await
        .map_err(BackendError::MoveExtractedFiles);

    if rename_result.is_err() {
        tokio::fs::remove_dir_all(project_folder)
            .await
            .map_err(BackendError::CleanExtractedFiles)?;
    }

    rename_result
}

/// resets the local data for the current project by removing the `.grafbase` directory
///
/// # Errors
///
/// - returns [`BackendError::ReadCurrentDirectory`] if the current directory cannot be read
///
/// - returns [`BackendError::DeleteDatabaseDirectory`] if the `.grafbase` directory cannot be deleted
pub fn reset() -> Result<(), BackendError> {
    let project = Project::get();

    fs::remove_dir_all(&project.database_directory_path).map_err(BackendError::DeleteDatabaseDirectory)?;

    Ok(())
}
