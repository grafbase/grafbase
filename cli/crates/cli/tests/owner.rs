#![allow(unused_crate_dependencies)]

use chrono::Duration;
use serde_json::json;

mod utils;

mod global {

    mod todo {
        use crate::utils::consts::{
            OWNER_TODO_CREATE, OWNER_TODO_DELETE, OWNER_TODO_GET, OWNER_TODO_LIST, OWNER_TODO_MIXED_SCHEMA,
            OWNER_TODO_OWNER_CREATE_SCHEMA, OWNER_TODO_SCHEMA, OWNER_TODO_UPDATE,
        };
        use crate::utils::environment::Environment;
        use crate::{admin_jwt, user_one_jwt, user_three_jwt, user_two_jwt};
        use backend::project::GraphType;
        use json_dotpath::DotPaths;
        use serde_json::{json, Value};

        #[ignore]
        #[test]
        fn entity_should_be_visible_only_to_the_owner() {
            let mut env = Environment::init();
            env.grafbase_init(GraphType::Single);
            env.write_schema(OWNER_TODO_SCHEMA);
            env.grafbase_dev();
            let client = env.create_client();
            client.poll_endpoint(30, 300);

            // user1 creates a todo.
            let todo_created = client
                .gql::<Value>(OWNER_TODO_CREATE)
                .bearer(&user_one_jwt())
                .variables(json!({ "title": "1", "complete": false }))
                .send();

            insta::assert_json_snapshot!("user1-create", todo_created, {".data.todoCreate.todo.id" => "[id]"});
            let id: String = todo_created
                .dot_get("data.todoCreate.todo.id")
                .unwrap()
                .expect("id must be present");
            // user1.list should see the todo.
            insta::assert_json_snapshot!(
                "user1-list",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_one_jwt()).send()
            );
            // user1 should be able to get the todo by id.
            insta::assert_json_snapshot!(
                "user1-get",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&user_one_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // user1 updates the todo.
            insta::assert_json_snapshot!(
                "user1-update",
                client
                    .gql::<Value>(OWNER_TODO_UPDATE)
                    .bearer(&user_one_jwt())
                    .variables(json!({"id": id, "input": { "complete": true }}))
                    .send()
            );
            // user1.list should see the todo with updated complete status.
            insta::assert_json_snapshot!(
                "user1-list-2",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_one_jwt()).send()
            );
            // user2.list should be empty.
            insta::assert_json_snapshot!(
                "list-empty",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_two_jwt()).send()
            );
            // user2 should not be able to get the todo by id.
            insta::assert_json_snapshot!(
                "user2-get-fail",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&user_two_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // an attempt by user2 to update the todo should fail.
            client
                .gql::<Value>(OWNER_TODO_UPDATE)
                .bearer(&user_two_jwt())
                .variables(json!({"id": id, "input": { "complete": false }}))
                .send();
            insta::assert_json_snapshot!(
                "user1-list-2",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_one_jwt()).send()
            );
            // an attemt by user2 to delete the todo should fail.
            client
                .gql::<Value>(OWNER_TODO_DELETE)
                .bearer(&user_two_jwt())
                .variables(json!({ "id": id }))
                .send();
            insta::assert_json_snapshot!(
                "user1-list-2",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_one_jwt()).send()
            );
            // user1 deletes the todo.
            insta::assert_json_snapshot!(
                "user1-delete",
                client
                    .gql::<Value>(OWNER_TODO_DELETE)
                    .bearer(&user_one_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // list of todos should be empty.
            insta::assert_json_snapshot!(
                "list-empty",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_one_jwt()).send()
            );
        }

        #[ignore]
        #[test]
        fn owner_create_group_all_should_work() {
            let mut env = Environment::init();
            env.grafbase_init(GraphType::Single);
            env.write_schema(OWNER_TODO_OWNER_CREATE_SCHEMA);
            env.grafbase_dev();
            let client = env.create_client();
            client.poll_endpoint(30, 300);

            // user1 creates a todo.
            let todo_created = client
                .gql::<Value>(OWNER_TODO_CREATE)
                .bearer(&user_one_jwt())
                .variables(json!({ "title": "1", "complete": false }))
                .send();
            insta::assert_json_snapshot!("user1-create", todo_created, {".data.todoCreate.todo.id" => "[id]"});
            let id: String = todo_created
                .dot_get("data.todoCreate.todo.id")
                .unwrap()
                .expect("id must be present");

            // // admin.list should see the todo.
            insta::assert_json_snapshot!("list", client.gql::<Value>(OWNER_TODO_LIST).bearer(&admin_jwt()).send());
            // user3.list should see the todo.
            insta::assert_json_snapshot!(
                "list",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_three_jwt()).send()
            );
            // user1.list should be unauthorized.
            insta::assert_json_snapshot!(
                "list-fail",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_one_jwt()).send()
            );
            // user2.list should be unauthorized.
            insta::assert_json_snapshot!(
                "list-fail",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_two_jwt()).send()
            );

            // user1 should be able to get the todo by id.
            insta::assert_json_snapshot!(
                "get",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&user_one_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // user2 should not be able to get the todo by id.
            insta::assert_json_snapshot!(
                "get-fail",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&user_two_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // admin should be able to get the todo by id.
            insta::assert_json_snapshot!(
                "get",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&admin_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // user3 should be able to get the todo by id.
            insta::assert_json_snapshot!(
                "get",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&user_three_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
        }

        #[ignore]
        #[test]
        fn group_should_supercede_owner_when_listing_entities() {
            let mut env = Environment::init();
            env.grafbase_init(GraphType::Single);
            env.write_schema(OWNER_TODO_MIXED_SCHEMA);
            env.grafbase_dev();
            let client = env.create_client();
            client.poll_endpoint(30, 300);

            // user1 creates a todo.
            let todo_created = client
                .gql::<Value>(OWNER_TODO_CREATE)
                .bearer(&user_one_jwt())
                .variables(json!({ "title": "1", "complete": false }))
                .send();
            insta::assert_json_snapshot!("user1-create", todo_created, {".data.todoCreate.todo.id" => "[id]"});
            let id: String = todo_created
                .dot_get("data.todoCreate.todo.id")
                .unwrap()
                .expect("id must be present");

            // admin.list should see the todo.
            insta::assert_json_snapshot!("list", client.gql::<Value>(OWNER_TODO_LIST).bearer(&admin_jwt()).send());
            // user3.list should see the todo.
            insta::assert_json_snapshot!(
                "list",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_three_jwt()).send()
            );
            // user1.list should see the todo.
            insta::assert_json_snapshot!(
                "list",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_one_jwt()).send()
            );
            // user2.list should be unauthorized.
            insta::assert_json_snapshot!(
                "list",
                client.gql::<Value>(OWNER_TODO_LIST).bearer(&user_one_jwt()).send()
            );

            // user1 should see the todo.
            insta::assert_json_snapshot!(
                "get",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&user_one_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // user2 should not be able to get the todo by id.
            insta::assert_json_snapshot!(
                "get-fail",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&user_two_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // admin should see the todo.
            insta::assert_json_snapshot!(
                "get",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&admin_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // user3 should see the todo.
            insta::assert_json_snapshot!(
                "get",
                client
                    .gql::<Value>(OWNER_TODO_GET)
                    .bearer(&user_three_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
        }
    }

    mod twitter {
        use crate::utils::consts::{
            OWNER_TWITTER_SCHEMA, OWNER_TWITTER_TWEET_CREATE, OWNER_TWITTER_USER_AND_TWEETS_GET_BY_ID,
            OWNER_TWITTER_USER_CREATE, OWNER_TWITTER_USER_GET_BY_EMAIL, OWNER_TWITTER_USER_GET_BY_ID,
        };
        use crate::utils::environment::Environment;
        use crate::{user_one_jwt, user_two_jwt};
        use backend::project::GraphType;
        use json_dotpath::DotPaths;
        use serde_json::{json, Value};

        #[ignore]
        #[test]
        fn get_by_id_should_be_filtered_by_the_owner() {
            let mut env = Environment::init();
            env.grafbase_init(GraphType::Single);
            env.write_schema(OWNER_TWITTER_SCHEMA);
            env.grafbase_dev();
            let client = env.create_client();
            client.poll_endpoint(30, 300);

            // user1 creates a user entity.
            let email: &str = "user1@example.com";
            let user_created = client
                .gql::<Value>(OWNER_TWITTER_USER_CREATE)
                .bearer(&user_one_jwt())
                .variables(
                    json!({ "username": "user1", "email": email, "avatar": "http://example.com", "url": "http://example.com" }),
                )
                .send();

            insta::assert_json_snapshot!("user1-create", user_created, {".data.userCreate.user.id" => "[id]"});
            let id: String = user_created
                .dot_get("data.userCreate.user.id")
                .unwrap()
                .expect("id must be present");
            // user1 can use get by id
            insta::assert_json_snapshot!(
                "user1-get",
                client
                    .gql::<Value>(OWNER_TWITTER_USER_GET_BY_ID)
                    .bearer(&user_one_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
            // user2 cannot get the user entity by id
            insta::assert_json_snapshot!(
                "user2-get-empty",
                client
                    .gql::<Value>(OWNER_TWITTER_USER_GET_BY_ID)
                    .bearer(&user_two_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
        }

        #[ignore]
        #[test]
        fn get_by_email_should_be_filtered_by_the_owner() {
            let mut env = Environment::init();
            env.grafbase_init(GraphType::Single);
            env.write_schema(OWNER_TWITTER_SCHEMA);
            env.grafbase_dev();
            let client = env.create_client();
            client.poll_endpoint(30, 300);

            // user1 creates a user entity.
            let email: &str = "user1@example.com";
            let user_created = client
                .gql::<Value>(OWNER_TWITTER_USER_CREATE)
                .bearer(&user_one_jwt())
                .variables(
                    json!({ "username": "user1", "email": email, "avatar": "http://example.com", "url": "http://example.com" }),
                )
                .send();

            insta::assert_json_snapshot!("user1-create", user_created, {".data.userCreate.user.id" => "[id]"});
            // user1 can create a tweet
            insta::assert_json_snapshot!(
                "user1-get",
                client
                    .gql::<Value>(OWNER_TWITTER_USER_GET_BY_EMAIL)
                    .bearer(&user_one_jwt())
                    .variables(json!({ "email": email }))
                    .send()
            );
            // user2 cannot get the user entity by email
            insta::assert_json_snapshot!(
                "user2-get-empty",
                client
                    .gql::<Value>(OWNER_TWITTER_USER_GET_BY_EMAIL)
                    .bearer(&user_two_jwt())
                    .variables(json!({ "email": email }))
                    .send()
            );
        }

        #[ignore]
        #[test]
        fn test_linking() {
            let mut env = Environment::init();
            env.grafbase_init(GraphType::Single);
            env.write_schema(OWNER_TWITTER_SCHEMA);
            env.grafbase_dev();
            let client = env.create_client();
            client.poll_endpoint(30, 300);

            // user1 creates a user entity.
            let email: &str = "user1@example.com";
            let user_created = client
                .gql::<Value>(OWNER_TWITTER_USER_CREATE)
                .bearer(&user_one_jwt())
                .variables(
                    json!({ "username": "user1", "email": email, "avatar": "http://example.com", "url": "http://example.com" }),
                )
                .send();

            insta::assert_json_snapshot!("user1-create", user_created, {".data.userCreate.user.id" => "[id]"});
            let id: String = user_created
                .dot_get("data.userCreate.user.id")
                .unwrap()
                .expect("id must be present");
            // user1 can create a tweet linked to the user entity
            insta::assert_json_snapshot!(
                "user1-create-tweet",
                client
                    .gql::<Value>(OWNER_TWITTER_TWEET_CREATE)
                    .bearer(&user_one_jwt())
                    .variables(json!({ "userId": id }))
                    .send(),
                {".data.tweetCreate.tweet.id" => "[id]"}
            );
            // user2 cannot get the entity by id
            insta::assert_json_snapshot!(
                "user2-create-tweet-fail",
                client
                    .gql::<Value>(OWNER_TWITTER_TWEET_CREATE)
                    .bearer(&user_two_jwt())
                    .variables(json!({ "userId": id }))
                    .send()
            );
            // user1 can use get by id
            insta::assert_json_snapshot!(
                "user1-and-tweets-get",
                client
                    .gql::<Value>(OWNER_TWITTER_USER_AND_TWEETS_GET_BY_ID)
                    .bearer(&user_one_jwt())
                    .variables(json!({ "id": id }))
                    .send()
            );
        }
    }
}

fn user_one_jwt() -> String {
    make_jwt(&json!({
        "iss": "https://idp.example.com",
        "sub": "user1"
    }))
}

fn user_two_jwt() -> String {
    make_jwt(&json!({
        "iss": "https://idp.example.com",
        "sub": "user2"
    }))
}

fn user_three_jwt() -> String {
    make_jwt(&json!({
        "iss": "https://idp.example.com",
        "groups": [
            "admin"
        ],
        "sub": "user3",
    }))
}

fn admin_jwt() -> String {
    make_jwt(&json!({
        "iss": "https://idp.example.com",
        "groups": [
            "admin"
        ],
    }))
}

fn make_jwt(claims: &serde_json::Value) -> String {
    use jwt_compact::{
        alg::{Hs256, Hs256Key},
        AlgorithmExt, Claims, Header, TimeOptions,
    };

    let claims = Claims::new(claims).set_duration_and_issuance(&TimeOptions::default(), Duration::days(7));

    let key = Hs256Key::new(b"abc123");
    let header = Header::empty().with_key_id("my-key");

    Hs256.token(&header, &claims, &key).unwrap()
}
