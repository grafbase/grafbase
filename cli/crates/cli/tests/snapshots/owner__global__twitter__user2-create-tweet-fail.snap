---
source: crates/cli/tests/owner.rs
expression: "client.gql::<Value>(OWNER_TWITTER_TWEET_CREATE).bearer(USER2).variables(json!({\n                \"userId\" : id\n            })).send()"
---
{
  "data": {
    "tweetCreate": null
  },
  "errors": [
    {
      "message": "An issue happened while applying the transaction.",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "tweetCreate"
      ]
    }
  ]
}
